packages/core/src/provider.tsx

```diff
@@ -76,8 +81,11 @@ export function Web3ReactProvider<T extends BaseProvider = Web3Provider>({
   )
 }
 
-export function useWeb3React() {
-  const web3 = useContext(Web3Context)
+export function useWeb3React<T extends BaseProvider = Web3Provider>() {
+  // see the comment in Web3ReactProvider for an explanation of the below
+  const web3 = useContext<Web3ContextType<T> | undefined>(
+    Web3Context as unknown as Context<Web3ContextType<T> | undefined>
```

Comment:
You shouldn't need to cast to unknown first, and the context should be inferrable:

```
const web3 = useContext(Web3Context as Context<Web3ContextType<T> | undefined>)
```